### PR TITLE
Bugfix/19224-treegraph-overlapping-labels-not-working

### LINKED ISSUE #19224: Treegraph dataLabels allowOverlap: false not working
#### Expected behaviour
When setting `treegraph.dataLabels.allowOverlap: false`, the labels should disappear when overlapping.

#### Actual behaviour
When setting `treegraph.dataLabels.allowOverlap: false`, the labels don't disappear when overlapping.

#### Live demo with steps to reproduce
https://jsfiddle.net/BlackLabel/t1kwr5b4/

#### Product version
Highcharts v11.1.0

#### Affected browser(s)
All

### PR DESCRIPTION
**Pull request by MarkusBarstad**
Fixed #19224, treegraph data labels did not hide when `allowOverlap` was `false`. Implemented overlap detection for labels with text path.

#### Upgrade note
Text path support is no longer part of the `highcharts.js` main bundle, but requires loading the `modules/textpath.js` bundle. It is however built into those series that use text paths by default.

___
For any HC pros reading this:
[This line](https://github.com/highcharts/highcharts/blob/a886783b907b4c980ce1e1220f9e5fe7ba1054c7/ts/Series/Treegraph/TreegraphSeries.ts#L520) confuses me.

We `splat` `this.options.dataLabels` into an array but we never use it again (except in this branch) 😕  To me it seems like this does nothing. Is there some reason not apparent in the function for doing this?

I am unsure on whether to remove this line and exactly the proper way to refer to options of dataLabels for the treegraph series.

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1204915984120495

### CODE REVIEW

File: ts/Series/Treegraph/TreegraphSeries.ts [hunk 1]
```diff
@@ -523,6 +524,29 @@ class TreegraphSeries extends TreemapSeries {
...
+                    }
+
+                    return labels;
+                };
+
+                this.chart.hideOverlappingLabels([
+                    ...getLabels(this.points),
+                    ...getLabels(this.links)
+                ]);
+            }
```
Comment (hubertkozik) [inline ts/Series/Treegraph/TreegraphSeries.ts:549]: I am not sure if that's a perfect solution for that issue. I think it would be much cleaner if we add only one line here: https://github.com/highcharts/highcharts/blob/ce0d44730018c8b6a9eceb0b081ba1c43f05419b/ts/Extensions/OverlappingDataLabels.ts#L329
something like `push(series.links || []);`.

File: ts/Series/Treegraph/TreegraphSeriesDefaults.ts [hunk 2]
```diff
@@ -228,6 +228,7 @@ const TreegraphSeriesDefaults = {
      */
     dataLabels: {
         defer: true,
+        allowOverlap: false,
```
Comment (hubertkozik) [inline ts/Series/Treegraph/TreegraphSeriesDefaults.ts:231]: I don't think we need this line to add, as Treegraph series is inheriting this option and API states correctly as `false`: https://api.highcharts.com/highcharts/series.treegraph.dataLabels.allowOverlap

**Review by hubertkozik — CHANGES_REQUESTED** [review]
Added comments below, but I don't think so, that in the current stage of your changes: https://jsfiddle.net/BlackLabel/kh7n10ry/ this issue is fixed.

To me, those links data labels are still overlapping with each other and they shouldn't. Currently, we don't have logic to figure out the box of data label with text path, so I think it would be necessary to implement it.

File: ts/Series/Treegraph/TreegraphSeries.ts (hunk 1 above)
Comment (MarkusBarstad) [reply ts/Series/Treegraph/TreegraphSeries.ts:549]: Ah, I can try to rewrite it to something like that!

File: ts/Series/Treegraph/TreegraphSeriesDefaults.ts (hunk 2 above)
Comment (MarkusBarstad) [reply ts/Series/Treegraph/TreegraphSeriesDefaults.ts:231]: For some reason I struggled with reffering to it but I will check again with your suggestion!

**Comment by MarkusBarstad** [comment]
> To me, those links data labels are still overlapping with each other and they shouldn't. Currently, we don't have logic to figure out the box of data label with text path, so I think it would be necessary to implement it.
True, they should be hidden when there is any overlapping at all. 

Do you think it would be appropriate to alter the `chart.hideOverlappingLabels` function to handle this case as well? It hides some of the overlapping labels but it seems like it does not handle rotation well or something

**Review by Xillez — APPROVED** [review]
Looks good to me, although I agree with Huberts comments.

**Comment by hubertkozik** [comment]
I didn't test it, but I think `chart.hideOverlappingLabels` works well, but `getBBox` of text with `textPath` return box without `textPath`, so I would try to add logic to `getBBox` to include `textPath` property and calculate proper box of that element.

File: ts/Series/Treegraph/TreegraphSeriesDefaults.ts (hunk 2 above)
Comment (MarkusBarstad) [reply ts/Series/Treegraph/TreegraphSeriesDefaults.ts:231]: Your suggestion works, my bad!

File: ts/Core/Renderer/BBoxObject.d.ts [hunk 3]
```diff
@@ -28,6 +28,7 @@ export interface BBoxObject extends PositionObject, SizeObject {
     width: number;
     x: number;
     y: number;
+    poly?: { x: number, y: number }[]
```
Comment (TorsteinHonsi) [inline ts/Core/Renderer/BBoxObject.d.ts:31]: If we use tuples, we'll produce more terse code, and we'll be able to reuse existing functions from map geometry code.
```suggestion
    polygon?: [number, number][]
```

File: ts/Extensions/OverlappingDataLabels.ts [hunk 4]
```diff
@@ -89,7 +89,42 @@ function chartHideOverlappingLabels(
             box2.x + box2.width <= box1.x ||
             box2.y >= box1.y + box1.height ||
             box2.y + box2.height <= box1.y
-        );
+        ),
+        pointIsInPolygon = (p: Vec, polygon: Vec[]): boolean => {
```
Comment (TorsteinHonsi) [inline ts/Extensions/OverlappingDataLabels.ts:93]: Later, we can import this function from https://github.com/highcharts/highcharts/blob/master/ts/Maps/MapUtilities.ts#L49-L81.

File: ts/Extensions/OverlappingDataLabels.ts [hunk 5]
```diff
@@ -89,7 +89,42 @@ function chartHideOverlappingLabels(
...
+                        x1
+                    )
+                ) {
+                    inside = !inside;
+                }
+            }
+            return inside;
+        },
+        boxCheck = (box1: BBoxObject, box2: BBoxObject): boolean => {
+            const box1Poly = box1.poly,
```
Comment (TorsteinHonsi) [inline ts/Extensions/OverlappingDataLabels.ts:117]: A possible performance improvement here would be to first check if the rectangular bounding boxes overlap. If they don't, we don't need to loop over all the vectors to check of an exact overlap.

File: ts/Series/Treegraph/TreegraphSeries.ts (hunk 1 above)
Comment (MarkusBarstad) [reply ts/Series/Treegraph/TreegraphSeries.ts]: Its not exactly like that now, but it is way way simpler as it turns out only the link-labels was needed!

**Review by Xillez — APPROVED** [review]
lgtm 👍

**Review by TorsteinHonsi — CHANGES_REQUESTED** [review]
Thanks, this is a solid piece of work! I’ll do a more detailed review when time. It adds 0.5 kB of weight, so I’m still considering whether we should add it to the core or to a module, but right now I’m inclining towards inclusion. Although text paths are not a very much used feature, rotated labels are. And very little of the added code deals specifically with text paths. I’ll also go over and see if there is room for shrinking.

For now, can you take a look at the changes in the visual tests? Seems like some labels are hidden that shouldn't be.

File: ts/Core/Renderer/SVG/SVGElement.ts [hunk 6]
```diff
@@ -1612,13 +1614,23 @@ class SVGElement implements SVGElementLike {
         const x = Math.min(aX, bX, cX, dX),
             y = Math.min(aY, bY, cY, dY),
             boxWidth = Math.max(aX, bX, cX, dX) - x,
-            boxHeight = Math.max(aY, bY, cY, dY) - y;
+            boxHeight = Math.max(aY, bY, cY, dY) - y,
+            [xCoord, yCoord] = ['x', 'y'].map((str): number => {
+                const attr = this.attr(str);
+                return isNumber(attr) ? attr : 0;
+            });
```
Comment (TorsteinHonsi) [inline ts/Core/Renderer/SVG/SVGElement.ts:1621]: It shouldn't be necessary to adjust the coordinates like this. There seems to be something wrong with the corner coordinates in the first place.

Consider this demo (running on this PR branch, checked out locally, as of today): https://jsfiddle.net/highcharts/a2L30w9v/

Observe that
* For a `label`, the bounding box rectangle is wrong. For a `text`, the bounding box rectangle is correct. Which shows that your correction for the x and y position should be applied earlier, to the `aX`, `bX` etc. constants.
* The polygon is flipped.

Comment (TorsteinHonsi) [reply ts/Core/Renderer/SVG/SVGElement.ts:1621]: And to elaborate, note that the bounding box for labels is wrong in the master too: https://jsfiddle.net/highcharts/a2L30w9v/1/

We might consider that an issue separate from this PR, but in any case this PR depends on it. We won't get the polygons right unless we fix the bounding box. @MarkusBarstad if you see fit, you can create a new issue on https://jsfiddle.net/highcharts/a2L30w9v/1/ and work on that separately, that's up to you.

Comment (TorsteinHonsi) [reply ts/Core/Renderer/SVG/SVGElement.ts:1621]: PS: The bounding box problem for labels also apply without rotation: https://jsfiddle.net/highcharts/a2L30w9v/2/. So this is likely an old problem. I think it comes from the SVGLabel being positioned by setting translation to the `g` element. Changes may also require updates in the [overlapping functionality](https://github.com/highcharts/highcharts/blob/e7b36f31692df459748ac15ab25b66a50a12b974/ts/Maps/Projection.ts#L158).

Comment (MarkusBarstad) [reply ts/Core/Renderer/SVG/SVGElement.ts:1621]: Thanks for good feedback!

Comment (hubertkozik) [reply ts/Core/Renderer/SVG/SVGElement.ts:1621]: On the master branch, we've got logic to get a box of the rotated label: https://jsfiddle.net/BlackLabel/uLt7rn95/ It is positioned to (0,0) because that was needed in other methods in HC.

Comment (TorsteinHonsi) [reply ts/Core/Renderer/SVG/SVGElement.ts:1621]: I've got a PR ready for it here: https://github.com/highcharts/highcharts/pull/20641 . Can't see any problems with it now... https://jsfiddle.net/highcharts/wng69bsq/

**Review by TorsteinHonsi — CHANGES_REQUESTED** [review]
Thanks so far! It looks like the math checks out now, but it needs some more work:

This is how it looks when I visualize (inline) the polygons of [one of our demos](https://jsfiddle.net/gh/get/library/pure/highcharts/highcharts/tree/master/samples/highcharts/members/renderer-textpath/).

Skjermbilde 2024-02-27 kl  15 48 00

The major problems are
1. Polygons should work on `text` elements as well as `label`. Currently it sits only in the `SVGLabel` class, while text paths are applied to `text` (`SVGElement`) in general.
2. The actual rendered polygon around the green circle is jagged.

I think it would be a good idea if we added commented-out code directly inside `getBBox` to visualize both the polygon and the rectangular bBox. Makes debugging much easier. Remove this block and replace it with something similar in `getBBox`:

https://github.com/highcharts/highcharts/blob/79b2b241c4366c5a56434fd6460393f72b4948f2/ts/Core/Series/DataLabel.ts#L357-L376

**Comment by TorsteinHonsi** [comment]
I didn't actually test this, but I suspect the current top-bottom, top-bottom polygon definition may create some false negative hit detection when the corner of one polygon hits the gap of another. Photoshopped:

Skjermbilde 2024-02-28 kl  15 23 23 copy

The way to overcome that would be to instead of pushing top and bottom points to the polygon, you would push the top point and unshift the bottom point. Or vice versa.

**Comment by MarkusBarstad** [comment]
> The way to overcome that would be to instead of pushing top and bottom points to the polygon, you would push the top point and unshift the bottom point. Or vice versa. Then the vertices won't cross.

I do not think this matters for overlap-detection, because the vertices can be in any order for our polygon detection function.
Here is a stress test, 1000 points, a `dataLabel` for all of them, rotated differently:
Screenshot 2024-02-28 at 20 36 45


However, if we do as you suggest we could have an easier time drawing the polygon with SVG paths, which could prove easier for debugging in the future.

**Comment by TorsteinHonsi** [comment]
> Here is a stress test, 1000 points, a dataLabel for all of them, rotated differently:

Are those just rotation, or are they text paths? Because those are two different ways of computing the polygon. 

Here's my concern, expressed in code: https://jsfiddle.net/highcharts/1kL56s72/

The green lines are the "polygons". In the fiddle, I made the text overlap without any of the polygon verteces crossing each other, which I suspect would count for a false negative.

**Comment by MarkusBarstad** [comment]
> > Here is a stress test, 1000 points, a dataLabel for all of them, rotated differently:
> 
> Are those just rotation, or are they text paths? Because those are two different ways of computing the polygon.
> 
> Here's my concern, expressed in code: https://jsfiddle.net/highcharts/1kL56s72/
> 
> The green lines are the "polygons". In the fiddle, I made the text overlap without any of the polygon verteces crossing each other, which I suspect would count for a false negative.

**Comment by MarkusBarstad** [comment]
Alright, @TorsteinHonsi , seems like a hit is detected in your demo! https://jsfiddle.net/mkb93/xaL6n5zw/86/

This is the exact same function as the one we use.

**Comment by TorsteinHonsi** [comment]
Thanks! Here's a false positive then: https://jsfiddle.net/highcharts/7852fu9y/ . It probably falsely detects an overlap because text 2 is surrounded by text 1.

I still think this would be avoided if the polygon was a nice fitting line around the text. Maybe the overlapping test must be updated too. It should only report a hit if any two _edges_ of the polygons are crossing.

**Comment by TorsteinHonsi** [comment]
Here's what I meant with part pushing, part unshifting the points:

```diff
diff --git a/ts/Core/Renderer/SVG/SVGElement.ts b/ts/Core/Renderer/SVG/SVGElement.ts
index 135d99f851..939f6fb575 100644
--- a/ts/Core/Renderer/SVG/SVGElement.ts
+++ b/ts/Core/Renderer/SVG/SVGElement.ts
@@ -1572,10 +1572,8 @@ class SVGElement implements SVGElementLike {
                             } = tp.getExtentOfChar(i),
                             top = y1 + offsetY,
                             left = x1 + offsetX;
-                        polygon.push(
-                            [left, top],
-                            [left, top + height]
-                        );
+                        polygon.push([left, top]);
+                        polygon.unshift([left, top + height])
                     }
 
                     const { x, y, width, height } = tp.getExtentOfChar(len),
@@ -1583,10 +1581,11 @@ class SVGElement implements SVGElementLike {
                         rightTop = y + offsetY;
 
                     // End of the polygon (vertex order does not matter)
-                    polygon.push(
-                        [rightEdge, rightTop],
-                        [rightEdge, rightTop + height]
-                    );
+                    polygon.push([rightEdge, rightTop]);
+                    polygon.unshift([rightEdge, rightTop + height]);
+
+                    // Close it
+                    polygon.push(polygon[0].slice() as [number, number]);
 
                     bBox.polygon = polygon;
                 }
```

The result is a polygon that (roughly) follows the perimeter of the text:
Skjermbilde 2024-03-01 kl  07 23 28

With this change, the false positive no longer happens, and as far as I can see the overlap detection works correctly. A hit is detected once the polygons visually overlap.

**Comment by TorsteinHonsi** [comment]
> I wish I knew a cheap recipe for more stable polygons which could handle twisting paths better

Yes, the current algorithm is increasingly wrong the more rotated the glyph is. With vertical text it falls apart:

Skjermbilde 2024-03-01 kl  13 13 50

I think there's one thing we should try. There is a function `getStartPositionOfChar`. This returns the position of the left bottom (or to be precise, _baseline_) position of the character. Now if we combine that with [SVGRenderer.fontMetrics](https://github.com/highcharts/highcharts/blob/2a775dccfef6899a2d74fe9604593bc55e2a4919/ts/Core/Renderer/SVG/SVGRenderer.ts#L1735-L1756) and the `getRotationOfChar`, we should be able to compute the correct top position too. For the very last element we would use `getEndPositionOfChar` instead.

In order to find the true bottom point, apply the baseline returned from `fontMetrics`. 

If this works, the only case I don't see we've covered, is multiline text. It may work out of the box, or we may have to deal with it separately for example to make sure the counter is reset and includes the first and last glyph on each line.

**Comment by TorsteinHonsi** [comment]
@MarkusBarstad I see you ended up using the center of `getExtentOfChar` instead of `getStartPositionOfChar`. Did you try `getStartPositionOfChar`?

**Comment by MarkusBarstad** [comment]
> @MarkusBarstad I see you ended up using the center of `getExtentOfChar` instead of `getStartPositionOfChar`. Did you try `getStartPositionOfChar`?

Yes I tried it, but I had some issues with it:
Screenshot 2024-03-08 at 15 03 01

By using 'h' from fontMetrics as height and using the rest of `getExtentOfChar`, [is the best solution I have found](https://jsfiddle.net/mkb93/xaL6n5zw/312/):

**Comment by TorsteinHonsi** [comment]
> Yes I tried it, @TorsteinHonsi , but I had some issues with it

Okay, something probably went wrong. In the latest commit I applied it, it looks good now.

**Comment by MarkusBarstad** [comment]
Alright, @TorsteinHonsi 
I think I am approaching a fix for multi-line ones (fiddle of latest commit): https://jsfiddle.net/mkb93/xaL6n5zw/636/

**Comment by TorsteinHonsi** [comment]
Thanks, we're closing in on the point when I can't tear this apart anymore. I found two fails now:

1. Text on a curve.  This worked correctly before your last one or two commits: https://jsfiddle.net/highcharts/smgczhfy/. But fails now: https://jsfiddle.net/highcharts/smgczhfy/1/
2. Text with no `tspan` never worked because of the way the query selector is set up: https://jsfiddle.net/highcharts/smgczhfy/2/

Other than that, some things remain before we can merge this:
1. Tests. Without them, we risk creating regressions for things that used to work, like text on a curve above. 
2. Final decision on the refactor. As it stands, we are adding 6-700 kB of code into the `highcharts.js` core for a feature that is useful, but quite fringe. In fact, text paths are not used at all for any of the core features. It is only used by arc diagram, dependency wheel, network graph, packed bubble, sunburst and treegraph. We should consider moving the whole text path stuff into a separate module/composition that would be included by those series type bundles. It would probably save us 1-3 kB of the main bundle. The downside is that those 1-3 kB would be repeated for each the series-type bundles, but it's probably not likely that many of those are loaded in the same page.

**Comment by MarkusBarstad** [comment]
Thank you for good feedback, @TorsteinHonsi !

I only have one question:
To make a polygon surrounding the curved textPath in your fiddle, we have to increase number of character-positions read, right?

Now we read every 5th character, so the polygon does not take the 'T' in your example into account, thus cutting through it.

Should we try every 3rd character instead?

**Comment by TorsteinHonsi** [comment]
> To make a polygon surrounding the curved textPath in your fiddle, we have to increase number of character-positions read, right?

I don't think so. It did work [after my last commit](https://jsfiddle.net/highcharts/smgczhfy/), which also used a step of 5. I think this was a regression that was introduced somehow with the multiline support.

**Comment by TorsteinHonsi** [comment]
Not sure if I understand the question, but by definition all texts with a break also have nested tspans. [Only the oneliners have raw text](https://github.com/highcharts/highcharts/blob/e8b2bc9858f3d2ddf46a1ab9060eb0147c4d59b1/ts/Core/Renderer/SVG/TextBuilder.ts#L121-L134).

**Comment by MarkusBarstad** [comment]
@TorsteinHonsi 

Found an easy solution, should work now: https://jsfiddle.net/mkb93/94vm3er5/24/

Note that this expects text wrapped in `span`-tags or non-wrapped text. Are there other HTML-tags we should expect in the input?

Also, in regards to making this an extension, should it be part of `OverlappingLabels.ts` 🤔 ?

**Comment by TorsteinHonsi** [comment]
Thanks!

Remember to create unit tests as you go along though. Now that we're down to fixing edge cases, it's a high risk that one fix introduces failures in previous cases. If I were you I would stop here and create a battery of tests for the cases we have been through so far (text/label, wrapped/bare content, oneline/multiline).

> Also, in regards to making this an extension, should it be part of OverlappingLabels.ts?

No, that module is included in the core. I think an entire new module is needed for everything `textPath`.

Next egde case: [nested spans](https://jsfiddle.net/highcharts/d3g58f12/). Maybe not very relevant, but I think we should consider replacing [this block](https://github.com/highcharts/highcharts/pull/20493/commits/d7db7d77947651ae51d7dd5307e9eb68fb138e66#diff-f19e21aba508aa49343d67aec2cb57ffd8a04916c41ef8b3232e81b7dbeb1722R1560-R1566) with some generic code that does `text.querySelectorAll('tspan')`, then do the `lines` logic separately on all of those. I don't have a clear idea of it though. It may be a problem if a tspan has some raw text content, then a child tspan in addition. You could spend some time on it, but abort mission if it becomes too complicated.

**Comment by MarkusBarstad** [comment]
@TorsteinHonsi 

Alright, the extension is taking shape but there are some potential issues with it.

I cannot find a way to add polygons to `bBox` within the extension. Instead I have reluctantly imported textPath polygon-generation code to `SVGElement`, and I call that code from `getBBox`. The reason for this is that `bBox` is not defined yet for `SVGElement`s within the extension.

My concern is that now every `SVGElement` will now import this code. Is there a slicker way of handling this 🤔 ?

**Comment by TorsteinHonsi** [comment]
@MarkusBarstad 
> Is there a slicker way of handling this
Yes, we typically use either of two patterns for cases like this:

1. Event-driven. The `SVGElement.getBBox` function would fire an event and your composition would add a handler. In this case I think a generic `afterGetBBox` event would make sense, though I'm not sure if it should run before or after the cache section. Maybe inside the `if (!bBox)` block so that we don't end up running it more than we need.
2. An optional function. The composition would add a function `applyTextPathPolygon` directly to SVGElement. The `getBBox` function would call it conditionally `this?.applyTextPathPolygon()`.

I think I prefer the event approach because it doesn't leave traces of fringe functionality in the core, and it opens up for future extensions and workarounds.

**Comment by MarkusBarstad** [comment]
@TorsteinHonsi I have gone with the event-driven approach but [some changes](https://github.com/highcharts/highcharts/blob/bugfix/19224-treegraph-overlapping-labels-not-working/ts/Core/Renderer/SVG/SVGElement.ts#L1551) to `SVGElement.getBBox` were necessary to store polygons on the bbox, and also to [later reference it](https://github.com/highcharts/highcharts/blob/bugfix/19224-treegraph-overlapping-labels-not-working/ts/Core/Renderer/SVG/SVGElement.ts#L1551) in `OverlappingLabels.ts`

I am also having a separate issue where [my tests](https://github.com/highcharts/highcharts/blob/bugfix/19224-treegraph-overlapping-labels-not-working/samples/unit-tests/text-path/labels/demo.js) seem to work locally but fail on github 🤔 
Is there a more correct way of importing extensions into tests or testing extensions in general?

**Comment by TorsteinHonsi** [comment]
> I have gone with the event-driven approach but [some changes](https://github.com/highcharts/highcharts/blob/bugfix/19224-treegraph-overlapping-labels-not-working/ts/Core/Renderer/SVG/SVGElement.ts#L1551) to SVGElement.getBBox were necessary to store polygons on the bbox, and also to [later reference it](https://github.com/highcharts/highcharts/blob/bugfix/19224-treegraph-overlapping-labels-not-working/ts/Core/Renderer/SVG/SVGElement.ts#L1551) in OverlappingLabels.ts

We have other cases where the event handler modifies a property later to be returned. What we usually do in this case is to add it to event arguments instead of as a property on the context element. A nice and clean example of this can be seen in the [Axis.autoLabelAlign](https://github.com/highcharts/highcharts/blob/65f21ce4d87089812802d515c0a330ecf107e61f/ts/Core/Axis/Axis.ts#L2823-L2839) function, where the subscribers modify `e.align`, and the mother function eventually returns the modified `evt.align`.

**Comment by MarkusBarstad** [comment]
@TorsteinHonsi 

I try to supply the bBox as an argument to the event like so: `fireEvent(this, 'afterGetBBox', bBox);`
This works neatly except when I try to commit the change, the [event selection](https://github.com/highcharts/highcharts/blob/master/samples/unit-tests/chart/events-selection/demo.js) unit-test fails giving me a big chunk of output 🤔
```
Chrome Headless 123.0.6312.124 (Mac OS 10.15.7) ERROR
  Uncaught TypeError: Cannot read properties of undefined (reading 'min')
  at code/highcharts.src.js:37566:54
  
  TypeError: Cannot read properties of undefined (reading 'min')
      at Chart.transform (code/highcharts.src.js:37566:54)
      at Chart.<anonymous> (code/highcharts.src.js:37634:34)
      at fireEvent (code/highcharts.src.js:1786:33)
      at Chart.transform (code/highcharts.src.js:37628:25)
      at Pointer.drop (code/highcharts.src.js:26057:31)
      at Pointer.onDocumentMouseUp (code/highcharts.src.js:26633:23)
      at TestController.triggerEvent (test/test-controller.js:990:17)
      at TestController.mouseUp (test/test-controller.js:452:14)
      at TestController.pan (test/test-controller.js:575:14)
      at Object.<anonymous> (samples/unit-tests/chart/events-selection/demo.js:24:16)Chrome Headless 123.0.6312.124 (Mac OS 10.15.7): Executed 9 of 182 SUCCESS (0
Chrome Headless 123.0.6312.124 (Mac OS 10.15.7) Highcharts Selection event FAILED
        global failure: TypeError: Cannot read properties of undefined (reading 'min')
        TypeError: Cannot read properties of undefined (reading 'min')
            at Chart.transform (code/highcharts.src.js:37566:54)
            at Chart.<anonymous> (code/highcharts.src.js:37634:34)
            at fireEvent (code/highcharts.src.js:1786:33)
            at Chart.transform (code/highcharts.src.js:37628:25)
            at Pointer.drop (code/highcharts.src.js:26057:31)
            at Pointer.onDocumentMouseUp (code/highcharts.src.js:26633:23)
            at TestController.triggerEvent (test/test-controller.js:990:17)
            at TestController.mouseUp (test/test-controller.js:452:14)
            at TestController.pan (test/test-controller.js:575:14)
            at Object.<anonymous> (samples/unit-tests/chart/events-selection/demo.js:24:16)
``` 

I have comitted with `--no-verify` just to show what I am trying to do.

**Comment by TorsteinHonsi** [comment]
Did you try the event args pattern seen in the other use cases?

`const e = { bBox }`
...
`fireEvent(this, 'afterGetBBox', e)`
...
`return e.bBox`;

File: ts/Core/Series/DataLabel.ts [hunk 7]
```diff
@@ -31,6 +31,8 @@ import type SVGLabel from '../Renderer/SVG/SVGLabel';
 import type SVGPath from '../../Core/Renderer/SVG/SVGPath';
 import type AnimationOptions from '../Animation/AnimationOptions';
 
+import TextPath from '../../Extensions/TextPath.js';
```
Comment (TorsteinHonsi) [inline ts/Core/Series/DataLabel.ts:34]: By importing TextPath into DataLabel.ts, we still make it part of `highcharts.js`, which was what we wanted to avoid.

Instead, we need to use an event in DataLabel.ts that the TextPath module can subscribe to, so that the module is not a dependency.

**Comment by MarkusBarstad** [comment]
@TorsteinHonsi 

I think i messed it up the first time around, it works now as per your example!

I have also:
- Made sure `drawTextPath` now happens on an event basis and removed the import
- Imported `textPath` module for all relevant samples I could find
- Moved the `TextPathObject` interface into the extension (though to me it seems unused) 🤔 

Any ideas on what more we should move?

`git grep` for `textPath` reveals checks for `textPath` but no `textPath`-only functionality

**Comment by MarkusBarstad** [comment]
@TorsteinHonsi 
Alright so:
- lower polygon line [fixed](https://jsfiddle.net/mkb93/9gx0tz16/68/)
- script-tags are removed from the relevant demos (still in `highcharts/members/renderer-textpath`),
- extension is imported into the relevant series

File: ts/Series/ArcDiagram/ArcDiagramSeries.ts [hunk 8]
```diff
@@ -28,6 +28,9 @@ import Series from '../../Core/Series/Series.js';
 import SeriesRegistry from '../../Core/Series/SeriesRegistry.js';
 import SVGRenderer from '../../Core/Renderer/SVG/SVGRenderer.js';
 import U from '../../Core/Utilities.js';
+import Highcharts from '../../Core/Globals.js';
+import TextPath from '../../Extensions/TextPath.js';
+TextPath.compose((Highcharts as AnyRecord).SVGElement);
```
[the remark this answers is not in the document]
Comment (MarkusBarstad) [reply ts/Series/ArcDiagram/ArcDiagramSeries.ts]: Alright, done!

Those series which used `import type SVGElement` were necessary to change to just `import SVGElement` since we cannot import `SVGElement` twice

File: ts/Series/Treegraph/TreegraphSeries.ts [hunk 9]
```diff
@@ -55,6 +56,9 @@ import TreegraphLayout from './TreegraphLayout.js';
 import { TreegraphSeriesLevelOptions } from './TreegraphSeriesOptions.js';
 import TreegraphSeriesDefaults from './TreegraphSeriesDefaults.js';
 import TreemapPoint from '../Treemap/TreemapPoint.js';
+import SVGElement from '../../Core/Renderer/SVG/SVGElement.js';
+import TextPath from '../../Extensions/TextPath.js';
+TextPath.compose(SVGElement);
```
Comment (bre1470) [inline ts/Series/Treegraph/TreegraphSeries.ts:61]: This compose call should run automatically sole in the masters containing treegraph. Treegraph also needs a static class function `compose`, if this is essential for Treegraph.

File: ts/Extensions/TextPath.ts [hunk 10]
```diff
@@ -0,0 +1,345 @@
...
+            );
+        }
+    }
+}
+
+function compose(SVGElementClass: typeof SVGElement): void {
+    addEvent(SVGElementClass, 'afterGetBBox', setPolygon);
+    addEvent(SVGElementClass, 'beforeAddingDataLabel', drawTextPath);
+    SVGElementClass.prototype.setTextPath = setTextPath;
+}
```
Comment (bre1470) [inline ts/Extensions/TextPath.ts:339]: This lacks protection against double calls. Check for `setTextPath`.

**Review by bre1470 — CHANGES_REQUESTED** [review]
Great work!

Just some requests regarding composition...

File: ts/Series/Treegraph/TreegraphSeries.ts (hunk 9 above)
Comment (MarkusBarstad) [reply ts/Series/Treegraph/TreegraphSeries.ts:60]: Or are you saying alter `treegraph.src.ts` to call this static compose function?

Comment (bre1470) [reply ts/Series/Treegraph/TreegraphSeries.ts]: This is necessary for alle the series here. Right now this extension is not an extension. It is not compatible with async ES imports and shaked bundling and is bundled everywhere.

Things to consider:

* If these series require the text path feature, they have to provide a compose function taking `SVGElement` for modification. The question then is, why do we have it as an extension?

* If it is optional, a call in `textpath` master is enough and the series do not need their own `compose`.

* The related masters need to call the compose function(s) in every case. They are our official bundles.

* If you want a truly extension, the `textpath` master needs to be the only one composing it. The `textpath` master can be imported by every other master that should contain the composition.

Comment (TorsteinHonsi) [reply ts/Series/Treegraph/TreegraphSeries.ts]: Okay, maybe we should discuss this offline, but here are the facts/requirements:
* TextPath is an extension to the SVGElement, allowing users to add text paths to text elements.
* So save weight, we don't want it in the main bundle.
* We provide a module bundle, `module/textpath.js`, so that users can use it in custom rendering.
* Some series types use text paths by default. For those series types, we want to embed the TextPath capabilities in the series type bundle.

Comment (bre1470) [reply ts/Series/Treegraph/TreegraphSeries.ts]: The compose in all series should still go into master files. It is not expected to have changes to SVGElement just by importing a series type. If this happens async, users have no chance of control when what gets modified. So remove the compose call from the series class files and add an import of the textpath master to the series masters regarding affected series types.

Comment (TorsteinHonsi) [reply ts/Series/Treegraph/TreegraphSeries.ts]: > users have no chance of control when what gets modified

The thing is that text paths just don't work without the extension. The data labels will not render correctly. So I'm not convinced that this should go in the masters. It's not a _modification_ per se of the SVGElement, it is the addition of a feature. I would imagine that if an implementer loads the Treegraph module using ESM and the data labels are all over the place, that would be surprising. I wouldn't think of loading the extra TextPath module explicitly.

Comment (bre1470) [reply ts/Series/Treegraph/TreegraphSeries.ts]: I see, so it is a must have.

**Review by bre1470 — APPROVED** [review]

File: ts/Series/Treegraph/TreegraphSeries.ts (hunk 9 above)
Comment (MarkusBarstad) [reply ts/Series/Treegraph/TreegraphSeries.ts:60]: Ok so I treegraph series now calls `TextPath.compose` as it was originally

I tried to revert the commit but somehow ended up merging over it when pulling from master, hopefully nothing weird happened to the branch :grimacing:

**Review by hubertkozik — COMMENTED** [review]
1. I tried to check at first the OP demo (check if datalabels are properly hidden if overlapped): https://jsfiddle.net/BlackLabel/gcwt75oh/ and it seems to me, that overlapping is still not working? Is it the desired behaviour?
2. Don't we need any documentation files? I think there should be information, that user should use a `textpath` module for specific cases.

**Comment by TorsteinHonsi** [comment]
> Regarding documentation, the series which needs textpath is supposed to include the extension, so I am not sure how we should document it

@MarkusBarstad It shouldn't be necessary to document it. Text path support it built into treegraph, so this should just work. If you inspect [the bundle](https://github.highcharts.com/bugfix/19224-treegraph-overlapping-labels-not-working/modules/textpath.js), you'll see that the `setTextPath` function exists there. So the issue lies somewhere else. Somehow it is not picked up.

**Comment by MarkusBarstad** [comment]
> > Regarding documentation, the series which needs textpath is supposed to include the extension, so I am not sure how we should document it
> 
> @MarkusBarstad It shouldn't be necessary to document it. Text path support it built into treegraph, so this should just work. If you inspect [the bundle](https://github.highcharts.com/bugfix/19224-treegraph-overlapping-labels-not-working/modules/textpath.js), you'll see that the `setTextPath` function exists there. So the issue lies somewhere else. Somehow it is not picked up.

I think I have found the issue, @TorsteinHonsi and @hubertkozik :
Somehow the `BBox` caching made it so every polygon in [OverlappingLabels.ts](https://github.com/highcharts/highcharts/blob/b657caa0d3c1d4a76cad8d7bac03aef1b905c6ba/ts/Extensions/OverlappingDataLabels.ts#L158) had the same polygon.

It can be fixed simply with disabling caching for any `BBox` which has a polygon. Not sure if this is an ok solution though.

**Comment by hubertkozik** [comment]
@MarkusBarstad there is an error in headless tests, can you check it?

**Comment by MarkusBarstad** [comment]
Alright, it passes now! @hubertkozik 

I got tripped up by the fact that SVG gets different dimensions in the headless browsers causing different labels to be hidden on the server side 🤔 

This might be a more inefficient unit-test because it involves looping through the labels and just checking that some are hidden/visible but it works.

**Review by hubertkozik — CHANGES_REQUESTED** [review]
I'm really sorry @MarkusBarstad for another question/requested change, but I believe it should work with your changes in this PR.

The rotation origin is wrong for data labels in treegraph: https://jsfiddle.net/BlackLabel/s450832r/ To reproduce you can move the slider to left and right. 

In the current version of Highcharts rotation works correctly: https://jsfiddle.net/BlackLabel/v5g6kqnp/
